### ISSUE TITLE
Train, Val, Test dataset leakage

### ISSUE THREAD

**Issue by Big-Tree** [thuml/Autoformer#20]
Hi, thanks for making your code public!  It's great for the forecasting community and research.

I am unsure about how you create your train, val and test splits.  The splits seem to be overlapping with a length of `self.seq_len`, see the code snippets below.

https://github.com/thuml/Autoformer/blob/7f9ce3c58b73b4f4b1163e093f47550a5cdbc6d5/data_provider/data_loader.py#L237-L240
https://github.com/thuml/Autoformer/blob/7f9ce3c58b73b4f4b1163e093f47550a5cdbc6d5/data_provider/data_loader.py#L267-L268
If `self.seq_len` is longer than `self.pred_len` then sequences that the model is trained on will be present in the validation split.  Data used for validation would also be present in the test split.

**Comment by wuhaixu2016** [comment]
Hi, thanks for your interest.

(1) I think this dataset splitting is right. You can understand this in the following two views.

- From the deep learning view, for a forecasting task, the key point is to not leak future information. The overlapping only happens on the encoder input. Thus, although the train, validation and test sets have the overlap with length `self.seq_len`, they will not affect the model learning and evaluation.
- From the application view, say you only have `num_train+num_vali time` points. If you want to evaluate the model performance in the future num_test time points, you will have to refer to the past `self.seq_len` in `num_train+num_vali` time points to obtain the first `self.pred_len` forecasting in the future.

From the above two views, this splitting is reasonable.

(2) I don't understand the condition "If self.seq_len is longer than self.pred_len". Would you please explain it to me in more detail?

- Note that the dataset splitting is only about `self.seq_len`. The `self.pred_len` does not appear.
- In line 285 of `data_loader.py`, the dataset length is set to `len(self.data_x) - self.seq_len - self.pred_len + 1`.

Thus, I do not find any unreasonable design.

**Comment by wuhaixu2016** [comment]
Cool! I see what you mean. And I agree with you that if pred_len < seq_len, there is leakage (in your view). **But, I still think this design is right in the time series forecasting task.** You can refer to item (1) in my last answer. 

(1) Here is a detailed explanation. In our design, I would like to re-define the Train, Val, Test as follows:

```
Whole series:  --------------------------------------------------
Train:         -----------------------------------
Validation:                                       -----
Test:                                                  ----------
seq_len = 5
```

**The splitting of Val and Test is based on the timestamps to be predicted.** As common knowledge, Val and Test are to select and eval models. And the metrics (e.g., MSE) are only calculated on the prediction. There is no overlapping that can affect the validation and test processes. And during these two processes, the addition seq_len time points are used for the first pred_len time points in the corresponding set.

Also, **this design just matches the real-world applications well**. For example, If I have 10 days of data, I split them into 8 days for training and the last 2 days for validation. Then, to measure the model performance, I will test the model in the last 2 days (Val set). For future testing,  to predict the 11th day, I will have to use the last seq_len timepoints in the 10 days that I already have. 

(2) If I set the shifted by pred_len, there will be prediction gap. 

As your suggestion, the splitting will be (seq_len=5, pred_len=2):

```
Whole series:  --------------------------------------------------
Train:         -----------------------------------
Validation:                                     -------
Test:                                                ------------
Prediction gap:                                        ---
```

I cannot obtain the forecasting results in the 3-5 days in the Test set. This is the prediction gap. In the real applications, this is unacceptable.

**Comment by Big-Tree** [comment]
I see your point.  Without merging the splits with a length of seq_len, it would not be possible to forecast time steps immediately after a split.  But I am not sure this out weighs the potential downsides.

(1) I am confused when you say that there is no overlapping that would effect the validation and the the test process.  If pred_len < leq_len then time steps used as the encoder input in the validation will also be used as the encoder input for the test set.  There is leakage between all the splits.

(2) In an extreme scenario where the `seq_len > (pred_len + val_split_len/2)` there would be leakage between train and test!

(3) You argue that merging splits is better in real world applications.  Do you think from a research perspective a hard split with no merging would be better?  This would be more simple to implement and does not risk data leakage.

I think it is important for there to be a standard way of creating dataset splits in the research community as this would aid in comparing results.

**Comment by wuhaixu2016** [comment]
Hi, I see what you mean and thanks for your detailed clarification. 

But I insist that deep models are developed for real-world applications. And this splitting method exactly matches the real-world scenarios. And I think most people who have a forecasting application will follow this splitting method for validation and test (serving).

（1）I think the key point of a forecasting task is to check whether the future is leaked or not. Obviously, this splitting method is without future leakage.

（2）For validation and test processes, the MSE (or other metrics) are only calculated on the predicted results. And the prediction intervals are non-overlapped in the validation and test sets.

If we extremely follow the research perspective, the splitting along the chronological order should also be reconsidered, which may also involve the temporal domain shift. Then, the standard way will be the follows:

- Split the whole time series into non-overlapped sub-series.
- Randomly select the sub-series for the train and test and validation sets.

But I would like to say, no one will conduct this in the continuous-time series dataset. Maybe this situation is also caused by the dataset property. In standard image or video datasets, such as ImageNet or Something-something, the samples are independent.

Thus, I do not have a certain idea of which splitting is better. Maybe future work can dig into this topic.

**Comment by wuhaixu2016** [comment]
I would like to close this issue. If someone still has problems, it is free to reopen this.

**Comment by Big-Tree** [comment]
Thanks, it's interesting to get your opinion.

**Comment by Zero-coder** [comment]
Good discussion!

**Comment by Zero-coder** [comment]
Simple explaination for train、val、test:[https://github.com/zhouhaoyi/Informer2020/issues/164](url)
